Simplify main loop

gassmoeller opened the pull request:
Another consequence of #1210. When reviewing that pull request, I again realized how atrociously complicated our main loop has become over time. This is an effort to simplify things to let new users (and me) understand how this loop works. The improvement is especially obvious in the 'split' code view of Github. I moved all complicated conditions and commands that take several lines into separate helper functions. I am open for name suggestions for these functions, I feel like `time_for_...` might be misleading, but have not found a better name yet.
This PR should wait for #1210 to be merged, and I will take another look at it after that.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2393,72 +2393,32 @@ namespace aspect
-          }
+        // if requested output a summary of the current timing information
+        if (time_for_timing_output())
+          computing_timer.print_summary ();
```
in other places, we have combined these two functions into a function that would be called `maybe_print_timing_output()` and that internally does the test and if necessary calls `print_summary()`. This would get you down to just one line here.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2476,30 +2436,8 @@ namespace aspect
+        // Do a checkpoint if indicated by checkpoint parameters
+        if (time_for_checkpoint(last_checkpoint_time,termination))
           {
             create_snapshot();
```
same here -- this could be combined into a function `maybe_create_snapshot`.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2393,72 +2393,32 @@ namespace aspect
+        if (time_for_additional_refinement_cycle())
+          do_additional_refinement_cycle(max_refinement_level);
+        else if (time_for_regular_refinement_cycle())
+          refine_mesh (max_refinement_level);
```
like for the cases below, everything you have (the 4 lines here + what's in at least `do_additional_refinement_cycle`) here could probably be combined into a function `maybe_refine_mesh()`.

bangerth commented on include/aspect/simulator.h:
```diff
@@ -968,6 +968,44 @@ namespace aspect
       /**
+       * Check if an additional refinement cycle is required in this timestep.
+       */
+      bool time_for_additional_refinement_cycle () const;
```
I think we've been pretty good at annotating each function's documentation with the file in which it is implemented.

gassmoeller commented:
I rebased and addressed the comments. The main loop looks much nicer now. The addition of the time step to the statistics object should also move in some helper function, either into `start_timestep()`, or into `compute_timestep()`, but I will leave that for a follow-up PR by somebody else, when addressing #1214, because it will change all test results. None of the changes in this PR should have an effect on the tests.

bangerth commented on include/aspect/simulator.h:
```diff
@@ -978,7 +1026,7 @@ namespace aspect
        * <code>source/simulator/helper_functions.cc</code>.
        */
-      std::pair<double,bool> compute_time_step () const;
+      double compute_time_step () const;
```
You'll have to update/delete the comment describing what the second part of the returned `std::pair` is.

bangerth commented on include/aspect/simulator.h:
```diff
@@ -968,6 +968,54 @@ namespace aspect
+       * pre_refinement_step variable is invalidated, and this function will
+       * return without doing refinement.
+       * An initial refinement cycle is different from a regular one,
+       * because the timestep is not increased afterwards. Instead the same
```
maybe say "time is not increased" or "time step number is not incremented"

bangerth commented on source/simulator/core.cc:
```diff
@@ -2377,95 +2375,32 @@ namespace aspect
-        if ( timestep_number == 0 )
-          pre_refinement_step = std::numeric_limits<unsigned int>::max();
+        if ((timestep_number == 0) && (maybe_do_initial_refinement(max_refinement_level)))
+          goto start_time_iteration;
```
I find function calls in conditionals difficult to read and understand if they have side effects. I think it may be easier to read the code if it read
```
  if (timestep_number == 0)
  {
    const bool initial_refinement_done = maybe_do...(...);
    if (initial_refinement_done == true)
      goto ...;
  }
```
That's more code but may be easier to read.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2479,40 +2414,8 @@ namespace aspect
-            last_checkpoint_time = std::time(NULL);
-          }
+        if (maybe_write_checkpoint(last_checkpoint_time,termination))
+          last_checkpoint_time = std::time(NULL);
```
same here

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -286,7 +286,134 @@ namespace aspect
+      }
+    // see if this is a time step where regular refinement is requested
+    else if (
+      (timestep_number > 0
```
let's start the actual condition on the line with the `if` as we do in other places. if you want to offset for better readability, add a couple of spaces

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -286,7 +286,134 @@ namespace aspect
+      }
+  }
+
+  template <int dim>
```
two extra lines. same after this function

bangerth commented on source/simulator/helper_functions.cc:
```diff
@@ -409,7 +536,11 @@ namespace aspect
+    // for now the bool (convection/conduction dominated)
+    // is unused, will be added to statistics later
     bool convection_dominant = (min_convection_timestep < min_conduction_timestep);
+    (void) convection_dominant;
```
why do you need this?

gassmoeller commented:
I think I have addressed all comments. Ready for another look and possibly merge.
